### PR TITLE
Feature/sl 964/pessimistic locking

### PR DESCRIPTION
**Pull request by radoslawlandowskikainos**
**Before creating a pull request make sure that:**

- [ ] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)

Please remove this line and everything above and fill the following sections:

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/sandl.snlevents/service/UserTransactionService.java [hunk 1]
```diff
@@ -6,11 +6,13 @@
 import uk.gov.hmcts.reform.sandl.snlevents.model.db.UserTransactionData;
 import uk.gov.hmcts.reform.sandl.snlevents.model.usertransaction.UserTransactionRulesProcessingStatus;
 import uk.gov.hmcts.reform.sandl.snlevents.model.usertransaction.UserTransactionStatus;
+import uk.gov.hmcts.reform.sandl.snlevents.repository.db.UserTransactionDataRepository;
 import uk.gov.hmcts.reform.sandl.snlevents.repository.db.UserTransactionRepository;
 
 import java.io.IOException;
 import java.util.List;
 import java.util.UUID;
+import java.util.function.Function;
```
Comment (r-kruk-kainos) [inline src/main/java/uk/gov/hmcts/reform/sandl.snlevents/service/UserTransactionService.java]: unused import

**Comment by r-kruk-kainos** [comment]
Fix checkstyle please.

File: src/main/java/uk/gov/hmcts/reform/sandl.snlevents/controllers/HearingPartController.java [hunk 2]
```diff
@@ -71,8 +72,8 @@ public ResponseEntity assignHearingPartToSession(
         @PathVariable UUID hearingPartId,
         @RequestBody HearingPartSessionRelationship assignment) throws IOException {
 
-        HearingPart hearingPart = hearingPartService.assignHearingPartToSession(hearingPartId, assignment);
+        UserTransaction ut = hearingPartService.assignHearingPartToSession(hearingPartId, assignment);
```
Comment (kpakur) [inline src/main/java/uk/gov/hmcts/reform/sandl.snlevents/controllers/HearingPartController.java]: (minor) I would consider change of the method name so it;s more obvious is going with user transaction....

File: src/main/java/uk/gov/hmcts/reform/sandl.snlevents/service/HearingPartService.java [hunk 3]
```diff
@@ -36,21 +46,59 @@ public HearingPart save(HearingPart hearingPart) {
...
 
-    public HearingPart assignHearingPartToSession(UUID hearingPartId,
+    @Transactional
+    public UserTransaction saveWithTransaction(HearingPart hearingPart, UUID transactionId) {
+        HearingPart savedHearingPart = hearingPartRepository.save(hearingPart);
+
+        List<UserTransactionData> userTransactionDataList = new ArrayList<>();
+        userTransactionDataList.add(new UserTransactionData("hearingPart",
+                savedHearingPart.getId(),
+                null,
```
Comment (kpakur) [inline src/main/java/uk/gov/hmcts/reform/sandl.snlevents/service/HearingPartService.java]: required for update to be able to update to the before state

File: src/main/java/uk/gov/hmcts/reform/sandl.snlevents/service/HearingPartService.java [hunk 4]
```diff
@@ -36,21 +46,59 @@ public HearingPart save(HearingPart hearingPart) {
...
+                "update",
+                "update",
+                0)
+        );
```
Comment (kpakur) [inline src/main/java/uk/gov/hmcts/reform/sandl.snlevents/service/HearingPartService.java]: 2 sessions should be here as well as these are form business point of few also parts of the transaction

**Pull request merged**